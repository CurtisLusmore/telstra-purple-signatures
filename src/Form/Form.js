import React, { Component } from 'react';
import { omit, mapValues } from 'lodash/object';
import { camelCase } from 'lodash/string';

import './Form.scss';
import constants from '../constants';
import SignatureContainer from '../SignatureContainer';

class Form extends Component {
  static labels = {
    signatureTypes: 'Signature Type',
    name: 'Your Name:',
    title: 'Job Title:',
    email: 'Email:',
    phone: 'Phone: (Optional)',
    mobile: 'Mobile:',
    twitter: 'Twitter: (Optional)',
    qualifications: 'Qualifications: (Optional)'
  };

  constructor() {
    super();
    this.state = {
      sigType: false,
      inputs: {
        signatureTypes: {
          text: [
            { text: 'Readify', checked: true, type: 'readify' },
            { text: 'Managed Services', checked: false, type: 'support' },
            { text: 'BTS Digital', checked: false, type: 'bts' }
          ].filter(Boolean),
          order: 1
        },
        name: { text: '', order: 2 },
        title: { text: '', order: 3 },
        qualifications: { text: '', order: 4 },
        mobile: { text: '', order: 5 },
        email: { text: '', order: 6 },
        twitter: { text: '', order: 7 }
      }
    };
  }

  handleChange = (name, value) => {
    let inputs = this.state.inputs;
    const newState = {
      ...inputs,
      [name]: { text: value, order: inputs[name].order }
    };
    this.setState({ inputs: newState });
  };

  handleRadioChange = index => {
    const signatureTypes = this.state.inputs.signatureTypes;
    const newSignatureTypes = signatureTypes.text.map((sigObj, sigIndex) => ({
      ...sigObj,
      ...(sigIndex === index ? { checked: true } : { checked: false })
    }));
    const sigType = newSignatureTypes[index].type;

    const baseInputs =
      sigType === 'bts'
        ? {
            phone: {
              text: '',
              order: 4.5
            },
            ...this.state.inputs
          }
        : omit(this.state.inputs, ['phone']);
    this.setState({
      inputs: {
        ...baseInputs,
        signatureTypes: { ...signatureTypes, text: newSignatureTypes }
      },
      sigType: sigType
    });
  };

  inputHtml = (inputName, inputVal, placeholders) => {
    if (inputName === 'signatureTypes') {
      return (
        <div className="field is-narrow">
          <div className="control">
            {inputVal.map((inputObj, index) => {
              return (
                <label className="radio" key={index}>
                  <input
                    className="radioInput"
                    type="radio"
                    name="signatureTypeOpt"
                    onChange={() => this.handleRadioChange(index)}
                    checked={inputObj.checked}
                    id={camelCase(inputObj.text)}
                  />
                  {inputObj.text}
                </label>
              );
            })}
          </div>
        </div>
      );
    }
    return (
      <input
        className="input"
        placeholder={placeholders[inputName]}
        style={{ width: '300px' }}
        value={inputVal || ''}
        onChange={e => this.handleChange(inputName, e.target.value)}
      />
    );
  };

  renderHtmlForInputs = (formInputs, placeholders) => {
    return Object.entries(formInputs)
      .sort((a, b) => a[1].order - b[1].order)
      .map(obj => ({ key: obj[0], ...omit(obj[1], 'order') }))
      .map(inputObj => {
        const inputName = inputObj.key;
        return (
          <div className="field is-horizontal" key={inputName}>
            <div className="field-label is-normal">
              <label className="label field-label">
                {Form.labels[inputName]}
              </label>
            </div>
            <div className="field-body">
              <div className="control" key={inputName}>
                {this.inputHtml(inputName, inputObj.text, placeholders)}
              </div>
            </div>
          </div>
        );
      });
  };

  render() {
    const placeholders =
      constants[this.state.sigType === 'bts' ? 'btsDigital' : 'readify']
        .placeholders;
    const formInputs = this.state.inputs;

    const inputs = this.renderHtmlForInputs(formInputs, placeholders);

    const SignatureContainerProps = {
      placeholders,
      sigType: this.state.sigType,
      ...omit(mapValues(this.state.inputs, 'text'), 'signatureType')
    };

    return (
      <form>
        {inputs}
        <hr />
        <SignatureContainer {...SignatureContainerProps} />
      </form>
    );
  }
}

export default Form;
