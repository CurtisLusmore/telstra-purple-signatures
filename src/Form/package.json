{
  "main": "./Form.js",
  "name": "Form",
  "version": "0.0.0",
  "private": true
}
