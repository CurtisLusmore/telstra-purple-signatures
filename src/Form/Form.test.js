import React from 'react';
import { shallow } from 'enzyme';

import Form from './Form';

it('Renders', () => {
  const tree = shallow(<Form />);
  expect(tree).toMatchSnapshot();
});
