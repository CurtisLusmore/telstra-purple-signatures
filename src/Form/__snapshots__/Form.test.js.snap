// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
<form>
  <div
    className="field is-horizontal"
    key="signatureTypes"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Signature Type
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="signatureTypes"
      >
        <div
          className="field is-narrow"
        >
          <div
            className="control"
          >
            <label
              className="radio"
              key="0"
            >
              <input
                checked={true}
                className="radioInput"
                id="readify"
                name="signatureTypeOpt"
                onChange={[Function]}
                type="radio"
              />
              Readify
            </label>
            <label
              className="radio"
              key="1"
            >
              <input
                checked={false}
                className="radioInput"
                id="readifySupport"
                name="signatureTypeOpt"
                onChange={[Function]}
                type="radio"
              />
              Readify Support
            </label>
            <label
              className="radio"
              key="2"
            >
              <input
                checked={false}
                className="radioInput"
                id="btsDigital"
                name="signatureTypeOpt"
                onChange={[Function]}
                type="radio"
              />
              BTS Digital
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="name"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Your Name:
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="name"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Your Name"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="title"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Job Title:
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="title"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Job Title"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="qualifications"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Qualifications: (Optional)
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="qualifications"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Jedi Master | PSM I"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="mobile"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Mobile:
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="mobile"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="+61 111 222 333"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="email"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Email:
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="email"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="your.name@readify.net"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="field is-horizontal"
    key="twitter"
  >
    <div
      className="field-label is-normal"
    >
      <label
        className="label field-label"
      >
        Twitter: (Optional)
      </label>
    </div>
    <div
      className="field-body"
    >
      <div
        className="control"
        key="twitter"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="@myTwitter"
          style={
            Object {
              "width": "300px",
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <hr />
  <withProps(mapProps(SignatureContainer))
    email=""
    mobile=""
    name=""
    placeholders={
      Object {
        "email": "your.name@readify.net",
        "mobile": "+61 111 222 333",
        "name": "Your Name",
        "qualifications": "Jedi Master | PSM I",
        "title": "Job Title",
        "twitter": "@myTwitter",
      }
    }
    qualifications=""
    sigType={false}
    signatureTypes={
      Array [
        Object {
          "checked": true,
          "text": "Readify",
          "type": "readify",
        },
        Object {
          "checked": false,
          "text": "Readify Support",
          "type": "support",
        },
        Object {
          "checked": false,
          "text": "BTS Digital",
          "type": "bts",
        },
      ]
    }
    title=""
    twitter=""
  />
</form>
`;
